src/libcore/array.rs

```diff
@@ -406,6 +408,73 @@ mod impls_using_const_generics {
         }
     }
 
+    #[unstable(feature = "array_value_iter", issue = "0")]
+    impl<T, const N: usize> IntoIterator for [T; N]
+    where
+        [T; N]: LengthAtMost32,
+    {
+        type Item = T;
+        type IntoIter = IntoIter<T, {N}>;
+
+        fn into_iter(self) -> Self::IntoIter {
+            IntoIter {
+                data: ManuallyDrop::new(self),
+                pos: 0,
+            }
+        }
+    }
+
+    /// A by-value array iterator.
+    ///
+    /// This struct is created by the `IntoIterator::into_iter`] method on
+    /// arrays.
+    #[unstable(feature = "array_value_iter", issue = "0")]
+    pub struct IntoIter<T, const N: usize> {
+        /// This is the array we are iterating over. Elements that were already
+        /// yielded (the ones with index `< self.pos`) have been read via
+        /// `ptr::read` and must not be accessed anymore!
+        data: ManuallyDrop<[T; N]>,
+        pos: usize,
```

Comment:
You'll need an end index as well to support `DoubleEndedIterator`.

Reply:
What came of @alexcrichton's suggestion to use `Range<usize>` in https://github.com/rust-lang/rust/pull/49000#discussion_r175811943?

Reply:
Hmm, I don't think I ever got around to trying a `Range`.

FWIW, here's my initial attempt with const generics, but I think it wasn't working yet: https://github.com/rust-lang/rust/compare/master...cuviper:array-intoiter-const

Reply:
I used `Range<usize>` now.

PS: thanks @cuviper for the link. I used some of your unit tests!